src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+
+        [ImportingConstructor]
+        public TempPEBuildManager(IProjectThreadingService threadingService,
+            IUnconfiguredProjectCommonServices unconfiguredProjectServices,
+            ILanguageServiceHost languageServiceHost,
+            IActiveConfiguredProjectSubscriptionService projectSubscriptionService,
+            [Import(typeof(VSLangProj.BuildManager))]VSBuildManager buildManager,
+            ITempPECompiler compiler,
+            IFileSystem fileSystem)
+             : base(threadingService.JoinableTaskContext)
+        {
+            _unconfiguredProjectServices = unconfiguredProjectServices;
+            _languageServiceHost = languageServiceHost;
+            _projectSubscriptionService = projectSubscriptionService;
+            _buildManager = buildManager;
+            _compiler = compiler;
+            _fileSystem = fileSystem;
+        }
+
+        public string[] GetTempPESourceFileNames()
+        {
+            Initialize();
+
+            return AppliedValue.Value.Inputs.ToArray();
+        }
+
+        /// <summary>
+        /// Called externally to fire the TempPEDirty events if the provided fileName is one of the known TempPE inputs or shared inputs
+        /// </summary>
+        public async Task TryFireTempPEDirtyAsync(string fileName)
+        {
+            // This method gets called for any file change but unless someone has asked for TempPE information this
+            // class will be uninitialized, so just exit early
+            if (!IsInitialized) return;
+
+            DesignTimeInputsItem inputs = AppliedValue.Value;
+
+            // This might be overkill, but none of the default .NET Core templates have any shared design time files
+            // Non shared design time files will be more common (just needs one .resx files), but its still reasonable to assume these collections will be empty a lot of the time
+            // Since this method is called every time a file is saved checking Count saves us 54ns for the fast path, and only costs 1ns the rest of the time
+            if (inputs.SharedInputs.Count > 0 && inputs.SharedInputs.Contains(fileName))
+            {
+                foreach (string item in inputs.Inputs)
+                {
+                    await FireTempPEDirtyAsync(item, false);
+                }
+            }
+            else if (inputs.Inputs.Count > 0 && inputs.Inputs.Contains(fileName))
+            {
+                await FireTempPEDirtyAsync(fileName, true);
+            }
+        }
+
+        /// <summary>
+        /// Called privately to actually fire the TempPE events and optionally recompile the TempPE library for the specified inpu
+        /// </summary>
+        private async Task FireTempPEDirtyAsync(string fileName, bool shouldCompile)
+        {
+            // Not using use the ThreadingService property because unit tests
+            await _unconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            if (shouldCompile)
```

Comment:
Why do we fire OnDesignTimeOutputDirty when we didn't compile anything?

Reply:
Mainly to match legacy behaviour. VSTypeResolutionService gets the event, and if it asks for the TempPE info again we'll recompile then. Legacy doesn't immediately compile for any shared input changes so I mirrored that behaviour.

Reply:
Why would we want VSTypeResolutionService to do anything in response to this change if the dll we return hasn't changed?

Reply:
From reading the legacy code the only thing I can deduce is that it would have been done this way for performance. When a shared input changes it fires the dirty event for all inputs so recompiling all inputs could take time. If there isn't a designer open, then nobody is listening to the events, so a recompile will be delayed until the next time someone asks for a dll.

> the dll we return hasn't changed

When we get asked to return it, we check the file date and recompile it, so by the time it is returned it will have changed.